ultravox/tools/ds_tool.py

```diff
@@ -0,0 +1,172 @@
+
+Question: {question}
+
+Answer: {answer}
+
+Provide a short explanation to the question given the passage that entails the answer."""
+
+
+@dataclasses.dataclass(frozen=True)
+class TtsTask:
+    column_name: str = simple_parsing.field(default="question", alias="-c")
+    audio_column_name: Optional[str] = simple_parsing.field(default=None, alias="-a")
+    voice: Optional[str] = simple_parsing.field(default=None, alias="-V")
+    sample_rate: int = simple_parsing.field(default=16000, alias="-r")
+
+
+@dataclasses.dataclass(frozen=True)
+class TextGenerationTask:
+    language_model: str = simple_parsing.field(default="gpt-4o", alias="-m")
+    new_column_name: str = simple_parsing.field(default="explanation", alias="-c")
+    template: str = DEFAULT_TEXTGEN_TEMPLATE
+
+
+# This script is used to either generate audio samples from text using a TTS model, or to generate text samples using a text generation model.
+# Ex: just ds_tool -T tts -d google/boolq -u fixie-ai/boolq-audio -c question -a audio
+# Ex: just ds_tool -T textgen -d fixie-ai/boolq-audio -u fixie-ai/boolq-audio -c explanation
+# Ex: just ds_tool -T textgen -d ylacombe/expresso -u fixie-ai/expresso -c continuation \
+#         --template "\"Continue the following sentence in a way that reflects a ‘{style}’ tone in a coherent style:\n{text}\""
+@dataclasses.dataclass
+class DatasetToolArgs:
+    dataset_name: str = simple_parsing.field(alias="-d")
+    dataset_subset: str = simple_parsing.field(default="default", alias="-S")
+    dataset_split: Optional[str] = simple_parsing.field(default=None, alias="-s")
+
+    num_samples: Optional[int] = simple_parsing.field(default=None, alias="-n")
+    num_workers: int = simple_parsing.field(default=16, alias="-w")
+
+    upload_name: Optional[str] = simple_parsing.field(default=None, alias="-u")
+    upload_branch: Optional[str] = simple_parsing.field(default="main", alias="-b")
+    num_shards: Optional[int] = simple_parsing.field(default=None, alias="-N")
+
+    token: Optional[str] = simple_parsing.field(default=None, alias="-t")
+
+    task: Union[TtsTask, TextGenerationTask] = simple_parsing.subgroups(
+        {"tts": TtsTask, "textgen": TextGenerationTask},  # type: ignore
+        default_factory=TtsTask,
+        alias="-T",
+    )
+
+
+def _tts_sample(
+    sample, col_name: str, audio_col_name: str, voice: str, sample_rate: int
+):
+    text = sample[col_name]
+    text = text["text"] if isinstance(text, dict) else text
+    sample[audio_col_name] = TTS_CLIENT.tts(text, voice=voice, sample_rate=sample_rate)
+    return sample
+
+
+def _tts_split(
+    ds_split: datasets.Dataset, task: TtsTask, num_proc: Optional[int] = None
+):
+    col_name = task.column_name
+    audio_col_name = task.audio_column_name or f"{col_name}_audio"
+    print(f'TTS mapping "{col_name}" to "{audio_col_name}"...')
+
+    return ds_split.map(
+        _tts_sample,
+        num_proc=num_proc,
+        fn_kwargs={
```

Comment:
FWIW, I liked the local function approach that I had before as it avoided having to curry all these parameters to the map function

Reply:
I feel like that approach adds complexity without any value:
1. we needed to handle the threading ourselves, which is bug prone.
2. we needed to handle the batched version of the code which for us provided no value, just needless complexity (batching is faster when you do things like vector math, but we'd just be batching and unbatching needlessly).

Reply:
oh yeah super glad to see the threading done internally, just wondering if we can use a local func to avoid having to pass all these `fn_kwargs` since the local func will have access to the local vars directly

Reply:
I assumed nested functions were usually a no/no:
![image](https://github.com/fixie-ai/ultravox/assets/5062458/f21522f8-aa35-4fbb-b414-579181bfcca0)

Alternatively, I was going to give it just the task.
![image](https://github.com/fixie-ai/ultravox/assets/5062458/a0d46c37-f256-4289-b39f-9e8bf44d57df)

But if you don't see any issues with nested functions, then I can switch over.

Reply:
well, you're doing that exact thing - closing over a local value other than self/cls - so I think it's valid. That said I don't feel super strongly about this

Reply:
You're right, I read the "other than" incorrectly there.
Okay using nesting instead.

Reply:
See if it actually makes it simpler - the use of `task` definitely helps a lot already, so use your own judgement as to whether the local functions are actually still useful.

Reply:
Both look similar. Not a big deal either way it can be changed later on if need be.